Bluetooth: controller: Periodic Advertising Report generation

cvinayak opened the pull request:
Added implementation for LE Periodic Advertising Report
event generation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix Extended Advertising Report generation

Fix Extended Advertising Report generation that skipped
population of AD data when there was no extended header.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Periodic Advertising Report generation

Added implementation for LE Periodic Advertising Report
event generation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -3976,6 +3976,193 @@ static void le_per_adv_sync_established(struct pdu_data *pdu_data,
+	uint8_t data_len = 0U;
+	int8_t tx_pwr = 0x7f;
+	uint8_t *data = NULL;
+	int8_t rssi = 0x7f;
```
Since `rssi` is assigned in line 4011, isn't this initialization unnecessary?

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -3976,6 +3976,193 @@ static void le_per_adv_sync_established(struct pdu_data *pdu_data,
+		}
+
+		len = ptr - (uint8_t *)p;
+		uint8_t hdr_len = len - offsetof(struct pdu_adv_com_ext_adv,
```
Should this have be declaring at start of scope like the other variables?

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -3976,6 +3976,193 @@ static void le_per_adv_sync_established(struct pdu_data *pdu_data,
+		/* if data cannot fit the event, mark it as incomplete */
+		if (data_len > data_max_len) {
+			data_len = data_max_len;
+			data_status = BIT(0);
```
Could we add #defines for the data status values instead of `BIT(0)` and `BIT(1)`?

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -3976,6 +3976,193 @@ static void le_per_adv_sync_established(struct pdu_data *pdu_data,
+		}
+	} else {
+		/* Data incomplete and no more to come */
+		if ((tx_pwr == 0x7f) && !data) {
```
Seeing the value 0x7F for RSSI and Tx Power so many places, I'm surprised that we don't have a `#define` for that either. Something we could consider.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -1793,11 +1799,13 @@ static inline int rx_demux_rx(memq_link_t *link, struct node_rx_hdr *rx)
 #if defined(CONFIG_BT_OBSERVER)
 #if defined(CONFIG_BT_CTLR_ADV_EXT)
 	case NODE_RX_TYPE_EXT_1M_REPORT:
-	case NODE_RX_TYPE_EXT_2M_REPORT:
```
Why was only the 2M report removed here?

cvinayak replied:
yes, this is intentional, as 2M PHY is only used in auxiliary channels, lll_scan_aux does not generate `NODE_RX_TYPE_EXT_2M_REPORT`, but always generates `NODE_RX_TYPE_EXT_AUX_REPORT` (PHY value is read from the LLL context), and then the function `ull_scan_aux_setup` converts the node_rx to `NODE_RX_TYPE_EXT_2M_REPORT` if the auxiliary PDU was generated as part of Extended Scanning (and not a AUX_CHAIN_IND PDU of a Periodic Advertising Report).

Thalley commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -1808,22 +1816,7 @@ static inline int rx_demux_rx(memq_link_t *link, struct node_rx_hdr *rx)
-			phy = BIT(0);
-			break;
-		case NODE_RX_TYPE_EXT_2M_REPORT:
-			phy = BIT(1);
```
The assignment of `phy = BIT(0)` and `phy = BIT(2)` have been moved, but I don't see anything that sets `phy = BIT(1)` anymore; is that intentional?

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_scan_aux.c:
```diff
@@ -93,33 +93,79 @@ void ull_scan_aux_setup(memq_link_t *link, struct node_rx_hdr *rx, uint8_t phy)
+		aux = NULL;
 		scan = (void *)HDR_LLL2EVT(ftr->param);
+		sync = scan->per_scan.sync;
+		phy = BIT(0);
```
Sorry about all the macro suggestions, but I usually find macros easier to understand and read, than these literals: We could also add `define`s for the different PHYs

cvinayak committed:
Bluetooth: controller: Use HCI defines for constants

Use HCI defines for PHY, and RSSI constants.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: bsim: Periodic Advertising Report generation

Test the LE Periodic Advertising Report generation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

kruithofa commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c:
```diff
@@ -273,9 +273,28 @@ static void isr_rx(void *param)
+
+			ftr = &(node_rx->hdr.rx_ftr);
+			ftr->param = lll;
+			ftr->rssi = (rssi_ready) ? (radio_rssi_get() & 0x7f) :
```
Please replace the 0x7F with a define here as well

kruithofa commented on tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:
```diff
@@ -1120,6 +1132,46 @@ static void test_scanx_main(void)
+	}
+
+	printk("Waiting for Periodic Advertising Report of %u bytes...",
+	       sizeof(per_adv_data1));
```
per_adv_data1 (and per_adv_data2) are undefined. Is there a dependency on another PR?

cvinayak replied:
yes, depends on #29778

cvinayak commented:
@kruithofa please re-review your request should be addressed now.

wopu-ot commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4180,6 +4188,207 @@ static void le_per_adv_sync_established(struct pdu_data *pdu_data,
+	sep->length = data_len;
+	memcpy(&sep->data[0], data, data_len);
+
+le_per_adv_report_invalid:
+	/* Free the node_rx list */
+	node_rx_next = node_rx->hdr.rx_ftr.extra;
+	while (node_rx_next) {
+		node_rx_curr = node_rx_next;
+		node_rx_next = node_rx_curr->hdr.rx_ftr.extra;
+
+		node_rx_curr->hdr.next = NULL;
+		ll_rx_mem_release((void **)&node_rx_curr);
+	}
```
Please extract this into a separate function that can be shared with `le_ext_adv_report` and avoid the goto.

cvinayak replied:
done.

wopu-ot commented on tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:
```diff
@@ -1145,6 +1157,46 @@ static void test_scanx_main(void)
+	printk("Waiting for Periodic Advertising Report of %u bytes...",
+	       sizeof(per_adv_data1));
+	sync_report_len_prev = sync_report_len;
+	while (!is_sync_report || (sync_report_len == sync_report_len_prev)) {
+		is_sync_report = false;
+		k_sleep(K_MSEC(100));
+	}
```
Here and in other places I would prefer the pattern "sleep; if (cond) { ... }" instead of "while (cond) { ... sleep; }" to avoid hanging forever in case something does not work as expected.

cvinayak replied:
the k_sleep is added in the while loop to allow context switching in the native_posix/babblesim. If there is a failure, the babblesim simulation timeout will catch the hanging. Alternative is to use k_sem_take for each check, adding complexity to the callbacks before it can k_sem_give. But to keep the test simple and to be able to debug with printks in this loop for the multiple incoming reports, while loop seems suitable.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_scan_aux.c:
```diff
@@ -93,32 +93,78 @@ void ull_scan_aux_setup(memq_link_t *link, struct node_rx_hdr *rx, uint8_t phy)
-		scan = (void *)HDR_LLL2EVT(ftr->param);
-	}
+			lll_sync = ftr->param;
+			sync = (void *)HDR_LLL2EVT(lll_sync);
+
+			rx->handle = ull_sync_handle_get(sync);
```
Please avoid using `sync` as temporary variable here and use a separate auto variable.

cvinayak replied:
Done

cvinayak committed:
Bluetooth: controller: Fix extended adv/sync report buffer leak

Fix node rx buffer leak if extended advertising and/or
periodic sync reports are masked out by the host.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4180,6 +4188,207 @@ static void le_per_adv_sync_established(struct pdu_data *pdu_data,
+		uint8_t hdr_len;
+		uint8_t *ptr;
+
+		/* The Link Layer currently returns RSSI as an absolute value */
+		rssi = -(node_rx_curr->hdr.rx_ftr.rssi);
```
no strong opinion here, RSSI value is vendor dependent. A different PR to address any change makes sense.

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -3825,17 +3825,21 @@ static void le_ext_adv_report(struct pdu_data *pdu_data,
 		evt_type_curr = p->adv_mode;
 
 		if (!p->ext_hdr_len) {
+			hdr_len = ptr - (uint8_t *)p;
```
Fixed this to assign a constant value using `offsetof`

cvinayak committed:
Bluetooth: controller: Add code comments for ext adv report

Added code comments to explain the reasons to drop
incomplete advertising and sync report.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4021,11 +4026,14 @@ static void le_ext_adv_report(struct pdu_data *pdu_data,
 	} else {
 		/* Data incomplete and no more to come */
-		if (!(adv_addr || (adi && ((tx_pwr != 0x7f) || data)))) {
+		if (!(adv_addr ||
+		      (adi && ((tx_pwr != BT_HCI_LE_ADV_TX_POWER_NO_PREF) ||
+			       data)))) {
```
done

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4180,6 +4188,207 @@ static void le_per_adv_sync_established(struct pdu_data *pdu_data,
+		}
+	} else {
+		/* Data incomplete and no more to come */
+		if ((tx_pwr == BT_HCI_LE_ADV_TX_POWER_NO_PREF) && !data) {
```
done

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4180,6 +4188,207 @@ static void le_per_adv_sync_established(struct pdu_data *pdu_data,
+		}
+
+		if (!node_rx_next) {
+			if (sec_phy_curr) {
```
Done

cvinayak committed:
Bluetooth: controller: Minor readability different variable use

Use an explicitly named variable to convey auxiliary pointer
present in the PDU chain.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c:
```diff
@@ -344,8 +344,10 @@ int lll_adv_scan_req_report(struct lll_adv *lll, struct pdu_adv *pdu_adv_rx,
-						  0x7f;
+	node_rx->hdr.rx_ftr.rssi = (rssi_ready) ?
+				   (radio_rssi_get() &
+				    BT_HCI_LE_ADV_TX_POWER_NO_PREF) :
+				   BT_HCI_LE_ADV_TX_POWER_NO_PREF;
```
While the value of this is correct, I would prefer (for readability) a separate #define for the rssi.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:
```diff
@@ -1198,8 +1198,9 @@ static int isr_rx_scan_report(struct lll_scan *lll, uint8_t rssi_ready,
-				   (radio_rssi_get() & 0x7f)
-				   : 0x7f;
+				   (radio_rssi_get() &
+				    BT_HCI_LE_ADV_TX_POWER_NO_PREF) :
+				   BT_HCI_LE_ADV_TX_POWER_NO_PREF;
```
Ditto

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:
```diff
@@ -303,31 +303,17 @@ static int isr_rx_pdu(struct lll_scan_aux *lll, uint8_t rssi_ready)
 
-	ftr->rssi = (rssi_ready) ? (radio_rssi_get() & 0x7f) : 0x7f;
+	ftr->rssi = (rssi_ready) ? (radio_rssi_get() &
+				    BT_HCI_LE_ADV_TX_POWER_NO_PREF) :
+				   BT_HCI_LE_ADV_TX_POWER_NO_PREF;
```
Ditto

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c:
```diff
@@ -273,9 +273,30 @@ static void isr_rx(void *param)
+			ftr->param = lll;
+			ftr->rssi = (rssi_ready) ?
+				    (radio_rssi_get() &
+				     BT_HCI_LE_ADV_TX_POWER_NO_PREF) :
+				    BT_HCI_LE_ADV_TX_POWER_NO_PREF;
```
Ditto

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -806,9 +806,9 @@ uint8_t ll_adv_enable(uint8_t enable)
 #if defined(CONFIG_BT_CTLR_CONN_RSSI)
-		conn_lll->rssi_latest = 0x7F;
+		conn_lll->rssi_latest = BT_HCI_LE_ADV_TX_POWER_NO_PREF;
 #if defined(CONFIG_BT_CTLR_CONN_RSSI_EVENT)
-		conn_lll->rssi_reported = 0x7F;
+		conn_lll->rssi_reported = BT_HCI_LE_ADV_TX_POWER_NO_PREF;
```
Ditto

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_master.c:
```diff
@@ -189,9 +189,9 @@ uint8_t ll_create_connection(uint16_t scan_interval, uint16_t scan_window,
 #if defined(CONFIG_BT_CTLR_CONN_RSSI)
-	conn_lll->rssi_latest = 0x7F;
+	conn_lll->rssi_latest = BT_HCI_LE_ADV_TX_POWER_NO_PREF;
 #if defined(CONFIG_BT_CTLR_CONN_RSSI_EVENT)
-	conn_lll->rssi_reported = 0x7F;
+	conn_lll->rssi_reported = BT_HCI_LE_ADV_TX_POWER_NO_PREF;
```
Ditto

cvinayak committed:
Bluetooth: controller: Fix periodic sync conditional compilation

Fix Periodic Advertising Synchronization conditional
compilation to allow Extended Scanning feature without
Periodic Advertising Synchronization feature enabled.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>
